test/loading.jl

```diff
@@ -2,22 +2,51 @@
 
 using Base.Test
 
-@test @__LINE__ == 5
+# Tests for @__LINE__ inside and outside of macros
+@test (@__LINE__) == 6
+
+macro macro_caller_lineno()
+    @test 9 == (@__LINE__) != __source__.line > 12
+    return __source__.line
+end
+
+@test @macro_caller_lineno() == (@__LINE__) > 12
+
+# @__LINE__ in a macro expands to the location of the macrocall in the source
+# while __source__.line is the location of the macro caller
+macro nested_LINE_expansion()
+    return quote
+        return (@emit_LINE, $(__source__.line))
+    end
+end
+macro nested_LINE_expansion2()
+    return :((@emit_LINE, $(__source__.line)))
+end
+macro emit_LINE()
+    return quote
+        (@__LINE__, $(__source__.line))
+    end
+end
+@test (@emit_LINE) == ((@__LINE__) - 3, @__LINE__)
+@test @nested_LINE_expansion() == ((@__LINE__() - 4, @__LINE__() - 12), @__LINE__())
+@test @nested_LINE_expansion2() == ((@__LINE__() - 5, @__LINE__() - 9), @__LINE__())
```

Comment:
Ah, of course.  Thanks - this does make a lot of sense.  (The reason I ask is that I'm having odd LLVM compile errors trying to build julia-0.7 right now, so I haven't been able to just try the code myself.)

This seems to be both the benefit and downside of the solution here: a benefit, because the __source__ should be clearly visible and non-magical in the AST.  This is great.  A downside, because it'll disturb any code which processes `Expr(:macrocall, ...)` (which quite honestly probably isn't a whole lot of user code).